src/proc.cpp

```diff
@@ -789,7 +823,7 @@ bool terminal_give_to_job(const job_t *j, bool cont) {
     // http://curiousthing.org/sigttin-sigttou-deep-dive-linux In all cases, our goal here was just
     // to hand over control of the terminal to this process group, which is a no-op if it's already
     // been done.
-    if (tcgetpgrp(STDIN_FILENO) == j->pgid) {
+    if (j->pgid == -2 || tcgetpgrp(STDIN_FILENO) == j->pgid) {
         debug(4, L"Process group %d already has control of terminal\n", j->pgid);
```

Comment:
I'm not sure that's true. If a job was all functions or builtins that didn't fork, pgid should be `-2`, but since it's actually operating under the scope of the current shell instance, it technically has control of the shell, no?

Reply:
Shouldn't the pgid then be the pgid of the shell?

AFAIK, "-2" here is the sentinel value for "hasn't been set up yet".

Reply:
I'm pretty sure the code that sets the pgid isn't called since iirc it's basically set on the first forked command in the chain. If there was no pgid-related code, -2 is indeed a reflection on the fact that at no point was a pgid assigned.

I'd have to double-check, though.

Reply:
builtins that do not fork keep pgrp -2.